### ISSUE TITLE
Datatables output by the pretty formatter are displayed incorrectly by gradle

### ISSUE THREAD

**Issue by scottadav** [cucumber/cucumber-jvm#2536]
### 👓 What did you see?

Problem is:
- Data tables are displayed in console with excessive newlines

Defect occurs when:
- The tests contain a DataTable
- The tests are run using Gradle and the JUnit5 Platform Engine
- The pretty plugin is enabled
- Test logging is enabled

Here is an example of the console output from running a sample test with gradle:
```
$ ./gradlew test
Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: test

> Task :test

RunAllCucumberTests > Cucumber > Demonstrate datatable display bug > com.example.RunAllCucumberTests.Print datatable STANDARD_OUT

    Scenario: Print datatable   # features/DataTable.feature:3
      Given a data table exists # com.example.steps.Steps.aDataTableExists(io.cucumber.datatable.DataTable)
    
    | 
    ColumnA
     | 
    ColumnB
     | 
    ColumnC
     |

    
    | 
    1
 
 
 
 
 
 
     | 
    2
 
 
 
 
 
 
     | 
    3
 
 
 
 
 
 
     |

    
    | 
    4
 
 
 
 
 
 
     | 
    5
 
 
 
 
 
 
     | 
    6
 
 
 
 
 
 
     |

    
    | 
    7
 
 
 
 
 
 
     | 
    8
 
 
 
 
 
 
     | 
    9
 
 
 
 
 
 
     |

    
    | 
    *
 
 
 
 
 
 
     | 
    0
 
 
 
 
 
 
     | 
    #
 
 
 
 
 
 
     |


RunAllCucumberTests > Cucumber STANDARD_OUT

    1 Scenarios (1 passed)
    1 Steps (1 passed)
    0m0.146s


```

### ✅ What did you expect to see?

I expected to see this:

```
$ ./gradlew test
Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: test

> Task :test

RunAllCucumberTests > Cucumber > Demonstrate datatable display bug > com.example.RunAllCucumberTests.Print datatable STANDARD_OUT

    Scenario: Print datatable   # features/DataTable.feature:3
      Given a data table exists # com.example.steps.Steps.aDataTableExists(io.cucumber.datatable.DataTable)
        | ColumnA | ColumnB | ColumnC |
        | 1       | 2       | 3       |
        | 4       | 5       | 6       |
        | 7       | 8       | 9       |
        | *       | 0       | #       |
 
RunAllCucumberTests > Cucumber STANDARD_OUT

    1 Scenarios (1 passed)
    1 Steps (1 passed)
    0m0.146s
```

### 📦 Which tool/library version are you using?

Gradle 7.4.2 (but replicated on 6.9.2 as well)
Cucumber 7.3.2 (cucumber-java, cucumber-junit-platform-engine)
Junit Jupiter API 5.8.2
Junit Platform Suite 1.8.2
Java 11.0.15+0

See the following [gradle.lockfile](https://raw.githubusercontent.com/scottadav/cucumber-junit5-example/demonstrate-datatable-gradle-display-defect/gradle.lockfile)

### 🔬 How could we reproduce it?

Download the attached project, unzip, and run `./gradlew test`.

Download link:
https://github.com/scottadav/cucumber-junit5-example/archive/refs/heads/demonstrate-datatable-gradle-display-defect.zip

### 📚 Any additional context?

I loaded it into a debugger, and it seems to stem from the method `DataTableFormatter.formatTo(DataTable, Appendable)` making multiple calls to `appendable.append(CharSequence)`. It seems that gradle is adding a newline after each call to `append()`. 

This is of course a gradle problem, but they do not intend to fix it. There was already a [defect reported](https://issues.gradle.org/browse/GRADLE-3560?_ga=2.226194906.1595684626.1650929176-1318319819.1650316670) but it is marked as *Won't Fix*.

An easy workaround, however, would be to append the various strings to a `StringBuilder` and then make  a single call to `Appendable.append(CharSequence)`. 

There is already a [separate Cucumber issue](https://github.com/cucumber/cucumber-jvm/issues/2481) to make fewer calls to flush the appendable. While this isn't the exact same issue, it could relieve some pressure by making fewer calls to append.

----
 
*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/ISSUE_TEMPLATE/bug_report.md)*

**Comment by scottadav** [comment]
I couldn't find a way to attach a zip file, so instead here is a gitlab link to download the sample project from. This was taken from a contributed cucumber example project and modified to demonstrate the defect.

https://github.com/scottadav/cucumber-junit5-example/tree/demonstrate-datatable-gradle-display-defect

Once you download the project, just run `./gradlew test` to see the defect.

I use MacOS, but the same problem happens on our gitlab-ci build server (Linux).

**Comment by mpkorstanje** [comment]
> This is of course a gradle problem, but they do not intend to fix it. There was already a [defect reported](https://issues.gradle.org/browse/GRADLE-3560?_ga=2.226194906.1595684626.1650929176-1318319819.1650316670) but it is marked as Won't Fix.

I don't see the intend to not fix it reflected in the issue. It would appear that the issue was migrated and the original reporter abandoned it. Would you mind recreating the issue with Gradle?

**Comment by scottadav** [comment]
I was looking at this, from the linked gradle issue:
![image](https://user-images.githubusercontent.com/4095427/165821077-edad0e20-4f55-46ab-934a-d1d390e7e967.png)

**Comment by mpkorstanje** [comment]
I suspect that is the default close status of the issue. Reading the comments suggests that the issue should be reopened on Github.

**Comment by scottadav** [comment]
I opened a new defect for Gradle at https://github.com/gradle/gradle/issues/20599 . 

I noticed that Gradle seems to be adding a newline every time cucumber-jvm flushes the output stream. It is a common pattern with text output to flush only when a newline character is appended to the output stream. Do you accept pull requests? I tried to fix this in cucumber-jvm using 3 different methods, and they all worked and took me about 30 seconds each to implement.

Possible Fixes:
1. In DataTableFormatter.java, write to a StringBuilder and then call `appender.append(...)` only once for the entire DataTable. 
2. In NiceAppendable.java, only call `tryFlush()` if the text being appended ends in a newline character.
3. Stop flushing at all in NiceAppendable.java, and just use default buffering and flushing. 

In [issue 2481](https://github.com/cucumber/cucumber-jvm/issues/2481) it is reported that pretty print is slow because `Calls flush method on every printed line.`. It seems that flushing less frequently would solve both performance and the display defect. Is this something that could be considered?

Thanks!
